hal/spi_api.h

```diff
@@ -19,220 +18,305 @@
+    bool    msb_first;  /**< True to send/receive the most significant bit first. */
 
-/** Initialize the SPI peripheral
- *
- * Configures the pins used by SPI, sets a default format and frequency, and enables the peripheral
- * @param[out] obj  The SPI object to initialize
- * @param[in]  mosi The pin to use for MOSI
- * @param[in]  miso The pin to use for MISO
- * @param[in]  sclk The pin to use for SCLK
- * @param[in]  ssel The pin to use for SSEL
+    spi_mode_t  mode;   /**< Transmission mode. See spi_mode_t. */
+
+    PinName SS;         /**< Slave select pin. */
+    PinName MISO;
+    PinName MOSI;       /**< Might not be connected in 3-wire mode. */
+    PinName MCLK;
+
+    uint32_t fill_symbol; /**< only the n lower bits will be used. */
+    uint32_t clock_frequency; /**< MCLK frequency in Hz. */
+
+    bool clock_phase; /**< True if data line is valid when leaving active state. */
+    bool clock_polarity; /**< True if the clock's rest state is high (+Vcc). */
+    uint32_t word_length; /**< Length of a symbol in bit. */
+} spi_init_t;
+
+/**
+ * This enumerates the possible result of the spi_init function.
  */
-void spi_init(spi_t *obj, PinName mosi, PinName miso, PinName sclk, PinName ssel);
+typedef enum spi_result_t {
+    SPI_RESULT_OK, /**< Operation successful. */
+    SPI_RESULT_CONFIG_UNSUPPORTED, /**< The required parameters are not supported by the device. */
+    SPI_RESULT_INVALID_PARAM, /**< The given parameter(s) is/are invalid. */
+    SPI_RESULT_ALREADY_INITIALIZED, /**< The requested peripheral/SS pin is already initialized. */
+} spi_result_t;
 
-/** Release a SPI object
+/**
+ * This structure groups all required data to handle a SPI transaction.
  *
- * TODO: spi_free is currently unimplemented
- * This will require reference counting at the C++ level to be safe
+ * - If tx is NULL a fill symbol is sent instead. See spi_init_t.
+ * - If tx is not NULL, the tx buffer must be at least `tx_count` symbols long.
  *
- * Return the pins owned by the SPI object to their reset state
- * Disable the SPI peripheral
- * Disable the SPI clock
- * @param[in] obj The SPI object to deinitialize
- */
-void spi_free(spi_t *obj);
-
-/** Configure the SPI format
+ * - If rx is NULL, all read symbols are dropped.
+ * - If rx is not NULL, the rx buffer must be at least `rx_count` symbols long.
  *
- * Set the number of bits per frame, configure clock polarity and phase, shift order and master/slave mode.
- * The default bit order is MSB.
- * @param[in,out] obj   The SPI object to configure
- * @param[in]     bits  The number of bits per frame
- * @param[in]     mode  The SPI mode (clock polarity, phase, and shift direction)
- * @param[in]     slave Zero for master mode or non-zero for slave mode
+ * A symbol might be bigger than a byte. In such case symbols are read/written following the
+ * platform's endianness.
  */
-void spi_format(spi_t *obj, int bits, int mode, int slave);
+typedef struct spi_transfer_args_t {
```

Comment:
this first should be `spi_transfer_args_s`, however I see where this is going. But from the codebase, tags are named with suffix `_s`  (see _api headers for instance).

Reply:
Actually I switched to `_t` because doxygen was not linking them properly when I generated the doc locally. All references to `spi_transfer_args_t` would not work unless `spi_transfer_args_s` is used in documentation too, which I find misleading.
For consistency it is better to use either one or the other every where.
Also, if we decide to use `_s` for `typdef struct identifier_s identifier_t;`, for consistency, we would also be required to name enumerations with `_e` (or something similar) to distinguish `typedef enum identifier_e identifier_t;`.

Reply:
Looking at enums, seems like no tags there just a name.

The doxygen issue - was not aware of it. If we keep both the same or leave it as it is now in HAL, this should be probably captured in the code style.